reward/reward_distributor.go

```diff
@@ -38,113 +50,469 @@ type governanceHelper interface {
+	// after magma, burn half of gas
+	if rc.rules.IsMagma {
+		burnt := getBurnAmountMagma(rewardFee)
+		rewardFee = rewardFee.Sub(rewardFee, burnt)
+		burntFee = burntFee.Add(burntFee, burnt)
+	}
+
+	// after kore, burn fees up to proposer's minted reward
+	if rc.rules.IsKore {
+		burnt := getBurnAmountKore(rc, rewardFee)
+		rewardFee = rewardFee.Sub(rewardFee, burnt)
+		burntFee = burntFee.Add(burntFee, burnt)
+	}
 
-	tmpInt = tmpInt.Mul(blockReward, rewardConfig.pocRatio)
-	pocIncentive := big.NewInt(0).Div(tmpInt, rewardConfig.totalRatio)
+	logger.Debug("calcDeferredFee()",
+		"totalFee", totalFee.Uint64(),
+		"rewardFee", rewardFee.Uint64(),
+		"burntFee", burntFee.Uint64(),
+	)
+	return totalFee, rewardFee, burntFee
+}
 
-	tmpInt = tmpInt.Mul(blockReward, rewardConfig.kirRatio)
-	kirIncentive := big.NewInt(0).Div(tmpInt, rewardConfig.totalRatio)
+func getBurnAmountMagma(fee *big.Int) *big.Int {
+	return new(big.Int).Div(fee, big.NewInt(2))
+}
 
-	remaining := tmpInt.Sub(blockReward, cnReward)
-	remaining = tmpInt.Sub(remaining, pocIncentive)
-	remaining = tmpInt.Sub(remaining, kirIncentive)
-	pocIncentive = pocIncentive.Add(pocIncentive, remaining)
+func getBurnAmountKore(rc *rewardConfig, fee *big.Int) *big.Int {
+	cn, _, _ := splitByRatio(rc, rc.mintingAmount)
+	proposer, _ := splitByKip82Ratio(rc, cn)
 
-	// CN reward
-	b.AddBalance(proposer, cnReward)
+	logger.Debug("getBurnAmountKore()",
+		"fee", fee.Uint64(),
+		"proposer", proposer.Uint64(),
+	)
 
-	// Proposer gets PoC incentive and KIR incentive, if there is no PoC/KIR address.
-	// PoC
-	if common.EmptyAddress(pocAddr) {
-		pocAddr = proposer
+	if fee.Cmp(proposer) >= 0 {
+		return proposer
+	} else {
+		return new(big.Int).Set(fee) // return copy of the parameter
 	}
-	b.AddBalance(pocAddr, pocIncentive)
+}
+
+// calcSplit splits fee into (proposer, stakers, kgf, kir, reamining)
+// the sum of the output must be equal to (minted + fee)
+func calcSplit(rc *rewardConfig, minted, fee *big.Int) (*big.Int, *big.Int, *big.Int, *big.Int, *big.Int) {
+	totalResource := big.NewInt(0)
+	totalResource = totalResource.Add(minted, fee)
+
+	if rc.rules.IsKore {
+		cn, kgf, kir := splitByRatio(rc, minted)
+		proposer, stakers := splitByKip82Ratio(rc, cn)
+
+		proposer = proposer.Add(proposer, fee)
+
+		remaining := new(big.Int).Set(totalResource)
+		remaining = remaining.Sub(remaining, kgf)
+		remaining = remaining.Sub(remaining, kir)
+		remaining = remaining.Sub(remaining, proposer)
+		remaining = remaining.Sub(remaining, stakers)
+
+		logger.Debug("calcSplit() after kore",
+			"[in] minted", minted.Uint64(),
+			"[in] fee", fee.Uint64(),
+			"[out] proposer", proposer.Uint64(),
+			"[out] stakers", stakers.Uint64(),
+			"[out] kgf", kgf.Uint64(),
+			"[out] kir", kir.Uint64(),
+			"[out] remaining", remaining.Uint64(),
+		)
+		return proposer, stakers, kgf, kir, remaining
+	} else {
+		source := big.NewInt(0)
+		source = source.Add(minted, fee)
```

Comment:
@ian0371 is `source` different from `totalResource`?

Reply:
fixed in ce4e1013